src/tools/kata-ctl/src/version.rs.in

```diff
@@ -0,0 +1,14 @@
+// Copyright (c) 2020 Intel Corporation
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+//
+// WARNING: This file is auto-generated - DO NOT EDIT!
+//
+
+#![allow(dead_code)]
+
+pub const KATA_CTL_VERSION: &str = "@KATA_CTL_VERSION@";
```

Comment:
This is fine. However, there is a related problem:

```bash
# This is fine:
$ cargo run -- version
kata-ctl version 2.5.0-alpha2-4e2bca5c7e76edf27c28f52153cca089088c6705-dirty (type: rust)

# Not ok:
$ cargo run -- --version
$ kata-ctl 0.0.0
```

The version string for `--version` is coming from:

```rust
     let app = App::new(name)
          .version(crate_version!())  // <--- Here
```

And `crate_version!()` is getting the version string `0.0.0` from the `Cargo.toml` file. So we have "three" versions or version components:

- The overall Kata release version from the `VERSION` file.
- The git commit.
- The `kata-ctl` tool version (from `Cargo.toml`).

I think it makes sense to use all of these and since [semver](https://semver.org) can be abused to deal with this, we can do the following:

## 1. Update the crate version

Update the version in `Cargo.toml` to `0.0.1` (as `0.0.0` looks odd ;):

```diff
diff --git a/src/tools/kata-ctl/Cargo.toml b/src/tools/kata-ctl/Cargo.toml
index c3ec304ac..582ae631b 100644
--- a/src/tools/kata-ctl/Cargo.toml
+++ b/src/tools/kata-ctl/Cargo.toml
@@ -5,7 +5,7 @@

 [package]
 name = "kata-ctl"
-version = "0.0.0"
+version = "0.0.1"
 authors = ["The Kata Containers community <kata-dev@lists.katacontainers.io>"]
 edition = "2021"
```

## 2. Change `version.rs` to use a function

```diff
diff --git a/src/tools/kata-ctl/src/version.rs.in b/src/tools/kata-ctl/src/version.rs.in
index e33ddfbdd..77b79d035 100644
--- a/src/tools/kata-ctl/src/version.rs.in
+++ b/src/tools/kata-ctl/src/version.rs.in
@@ -7,8 +7,10 @@
 // WARNING: This file is auto-generated - DO NOT EDIT!
 //

-#![allow(dead_code)]
+use clap::crate_version;

-pub const KATA_CTL_VERSION: &str = "@KATA_CTL_VERSION@";
-pub const GIT_COMMIT: &str = "@COMMIT@";
+const KATA_CTL_VERSION: &str = "@KATA_CTL_VERSION@";

+pub fn get() -> String {
+    format!("{}-{}", KATA_CTL_VERSION, crate_version!())
+}
```
Note that by changing to a function, we get to make the auto-generated `const` value private (which is good), and we can also easily unit-test the `get()` function.

## 3. Change the code for the version option

```diff
@@ -20,57 +20,39 @@ const NAME: &str = "kata-ctl";
 fn real_main() -> Result<()> {
     let name = crate_name!();

+    let version = version::get();
+
     let app = App::new(name)
-        .version(crate_version!())
+        .version(&*version)
         .about(ABOUT_TEXT)
         .long_about(DESCRIPTION_TEXT)
```

## 4. Change the code for the version command

```diff
@@ -116,16 +96,11 @@ fn real_main() -> Result<()> {
             Ok(())
         }
         "version" => {
-            println!(
-                "{} version {} (type: rust)",
-                NAME,
-                version::KATA_CTL_VERSION,
-        );
+            println!("{} version {} (type: rust)", NAME, version);
             Ok(())
         }
```

---

If we do that, the version becomes something like this:

```
2.5.0-alpha2-4e2bca5c7e76edf27c28f52153cca089088c6705-dirty-0.0.1
\----------/ \--------------------------------------/ \---/ \---/
    (a)                          (b)                   (c)   (d)
```

Where,

- `(a)` is the Kata version.
- `(b)` is the Git commit.
- `(c)` is the branch state.
- `(d)` is the `kata-ctl` tool version.

Ideally we'd have the `dirty` branch state tag at the end of the string (aka switch (c) and (d) around), but I think we can live with that.

There is a tiny issue still with this:

| CLI | result |
|-|-|
| `--version` | `kata-ctl $version` |
| `version` | `kata-ctl version $version (type: rust)` |

However, I think we can live with this.

wdyt to all of the above @amshinde, @cmaf?